File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        dfs(root, ans);
        return ans;
    }


   int dfs(TreeNode* root, int& ans){
        if(!root) return 0;

        int l = max(dfs(root->left, ans), 0);
        int r = max(dfs(root->right, ans), 0);

        ans = max(ans, root->val+l+r);
        return root->val + max(l,r);
   }
};

File: Data Structures & Algorithms/count-subsequences/submission-1.java
class Solution {
    public int numDistinct(String s, String t) {
        int n = s.length(), m = t.length();
        int[][] dp = new  int[n][m];
        for(int[] d: dp) {
            Arrays.fill(d,-1);
        }
        return dfs(0,0,s,t, dp);
    }

    public int dfs(int i, int j, String s, String t, int[][] dp) {
        int n = s.length(), m = t.length();
        if(j>=m) {
            return 1;
        }
        if(i>=n){
            return 0;
        }

        if(dp[i][j] != -1) return dp[i][j];

        int ans = 0;

        if(s.charAt(i) == t.charAt(j)){
            ans+=dfs(i+1, j+1,s,t, dp);
        }
        ans += dfs(i+1, j, s, t, dp);

        dp[i][j] = ans;
        return ans;
    }
}
